### ISSUE TITLE
Proposed database schema

### ISSUE THREAD

**Issue by cgalbiati** [andreisarebi/civic-app#7]
This assumes firebase/nosql, but I made a few notes for how it would change in a relational system.  I added examples to the top few in addition to the 'types'
```
candidates: {
  [id]: {
    id: (uuid)
    name: (text)
    image: (url)
    party: (text)
    ...restOfInfo: ...
    responseIds?: [uuids of responses]
  },
  'c88d3aa0-6ea8-11e8-b566-0800200c9a66': {
    'id': 'c88d3aa0-6ea8-11e8-b566-0800200c9a66',
    'name': 'Obama',
    'image': 'some_url',
    ...
  }
}

questions: {
  [id]: {
    id: (uuid)
    text: (text)
    image: (url)
    category/topic: (text)
  },
  '8bb9a280-6ea8-11e8-b566-0800200c9a66': {
    'id': '8bb9a280-6ea8-11e8-b566-0800200c9a66',
    'text': 'Would you like a to have a superpower?',
    'category': 'awesomeness', (could be categories if we want to map to mult)
    'image': 'some_url',
  },
}

candidateResponses: {
  [id]: {
    id: (uuid)
    questionId: (id or pertaining question)
    candidateId: (id of pertaining candidate)
    questionResponse: (number)
    explanation: (text)
  },
  'e86e1650-6ea8-11e8-b566-0800200c9a66': {
    'id': 'e86e1650-6ea8-11e8-b566-0800200c9a66',
    'questionId': '8bb9a280-6ea8-11e8-b566-0800200c9a66',
    'candidateId': 'c88d3aa0-6ea8-11e8-b566-0800200c9a66',
    'questionResponse': 't'
    'explanation': 'Obama is awesome and thereby likes awesome things',
  },
}

users: {
  [id]: {
    id: (uuid)
    authUserId: (id from firebase auth - maybe would just be the same id)
    ...userInfo?: ...
    address: (text?)
    district: (text/uuid?)
    responseIds?: [uuids of responses]
    favorites: [uuids of candidates]
  }
}

userResponses: {
  [id]: {
    id: (uuid)
    questionId: (uuid of answered question)
    questionResponse: (? t/f? number?)
    importanceScale?: (number)
  }
}
```
Notes: 
* Nested lists of ids (favorites, responseIds) should probably actually be dictionaries according to firebase docs, even though they look like lists here
* If we use relational, tables would mostly look the same (collumns would be the props on the object), with the following exceptions
  * Several of the string enums (district, category, etc) would become their own tables, and the data stored on the question/user would just be the id
  * Favorites would be a join table

**Comment by gmferland** [comment]
That all looks great to me. Just a couple of small comments:

* On the userResponses/candidateResponses schema, I don't think we need a UUID for each response. Rather, I would include userId/candidateId there instead and the primary key (not sure how Firebase conceptualizes PK) would be userId/candidateId and questionId

* I would also like to see a timestamp on the userResponses/candidateResponses schema. That way we can support things like allowing people to take the survey multiple times and seeing how their responses change over time. Or, at the very least, we have a record of when someone first registered with the app (assuming you have to take the survey before doing any kind of meaningful interaction with the rest of the app)

**Comment by cgalbiati** [comment]
@gmferland how would you use candidate/user id as the pk on response, when a single user/candidate has responses to many questions? I was under the impression that pk's had to be unique (especially in firebase's json structure, where the pk is the prop of an object)

**Comment by gmferland** [comment]
Yeah, sorry if I misspoke but I meant to say the primary key would be the combination of userId/candidateId and questionId, assuming one response per person per question. Or, you could still have some sort of responseId that is unique and a primary key for that schema, but it doesn't have to be universally unique. My guess though if Firebase is at all like MongoDB is that it will automatically generate a UUID for every record. So it's not a big deal

**Comment by cgalbiati** [comment]
Yeah, I think generating universally unique keys is easier than thinking about whether they need to be.  CandidateId + questionId might allow us to look them up more easily, you would only have to look at the id, rather than the object to get where candidateId===x, but I think you would probably have to write your own query system to do so, and I'm pretty sure you would not be able to in firebase

**Comment by cgalbiati** [comment]
oops.  Did not mean to close.

Also documenting from eng huddle last night - instead of yes/no and importance, we are going to store a value from -5 (strong no) to +5 (strong yes)

**Comment by cgalbiati** [comment]
## Updated schema: 
- Storing candidate info in google sheets rather than firebase will reduce costs and eliminate the need to figure out how to transform it to json.
- Nesting user data under the user object in firebase will reduce calls, and we really need all the info all the time anyway

## Candidate data
- stored in google sheets
- This is using essentially a relational model, where main candidate data is in one table (sheet), and positions in another.  In a real relational schema, we would have a join tables to candidate/question/position, but instead, I have essentially indexed positions on candidateId, so when we bring them into the app, we can generate a nested object with each candidate's positions nested under their id, and each position keyed by question.
- Right now, this approach requires fetching all the data.  For the relatively small number of candidates, we are supporting for the MVP, I think this will not be a problem, but eventually, we will probably want to only fetch candidates a user can vote for, which will work a lot better in a schema that allows for WHERE, and we will probably want a real database
- More details (mostly for content) about how to enter data: https://docs.google.com/document/d/1f11QTuStFapPt3_aC0vO-Kxm9B9QU7x4mEl-5FI_QnM/edit
- This PR (https://github.com/civic-app/civic-app/pull/27) shows implementation
### Summaries: 
ex: https://docs.google.com/spreadsheets/d/1Enwx8oPTcGyhv_re0kk_F2QiK5V80Lg4OftBbmrotkQ/edit#gid=0
- Columns for each property ex: name, id, party pref, bio, etc
- Id to link candidates to positions

### Positions:
ex: https://docs.google.com/spreadsheets/d/1Kh36D5DOyO8_it63OD-o51b2F4rlKfeLwnCbnaQZ5XE/edit#gid=0
- columns for questionId, candidateId, response, explanation
- response is 1 for progressive, -1 for conservative, 0 for neutral/unknown

### Survey questions:
ex: https://docs.google.com/spreadsheets/d/1Kh36D5DOyO8_it63OD-o51b2F4rlKfeLwnCbnaQZ5XE/edit#gid=0
- columns for id, text, type, image

### Elections:
- Right now, just using the string names in the candidate data (ex: 'Governor', 'California District 39', etc)
- Eventually we might want to hold more data about an election (ex: what districts vote in it, how competitive, etc.  I imagine that would be another table, and should be able to be stored in a spreadsheet

## User data:
- Stored in firebase
- keyed by userId (the id generated by auth)
- contains info like id, district, address as keys
- Responses are nested as a key
  - user responses are a range from -2 to 2
- Favorites are nested as a key
  - candidates or events (to come later, as well as orgs, probably), are user `[id]: true` to store a favorite
- ex: 
```
{
  "users": {
    "Ra6l4NfjcLcc8XdP4gX1aWdhRRd2": {
      "id": "Ra6l4NfjcLcc8XdP4gX1aWdhRRd2",
      "address": "125 W 125th st, New York, NY, 10001",
      "district": "10",
      "responses": {
        "8bb9a280-6ea8-11e8-b566-0800200c9a66": {
          "questionId": "8bb9a280-6ea8-11e8-b566-0800200c9a66",
          "response": 2
        },
        "33f3204d0e5d4a69b614b97eb18b9b9c": {
          "questionId": "33f3204d0e5d4a69b614b97eb18b9b9c",
          "response": -1
        },
      },
      "favorites": {
        "candidates": {
          "c88d3aa0-6ea8-11e8-b566-0800200c9a66": true
        },
        "events": {}
      }
    }
  }
}
```
